### DISCUSSION TITLE
Opto input

### DISCUSSION THREAD

**Discussion by jonas1972** [hatlabs/discussions#12] [Q&A]
Hi,

I'm trying to use the opto input but are struggling. Shouldn't I be able to connect GND to the ISO_GND and either either GND or 3.3V (from the SH-ESP32 board) to ISO_IN and get the corresponding signal value on pin 35? Do I need any configuration in the software? The code is working just fine with the VN pin but if I change to pin 35, then I don't get any value change when changing the ISO_IN input.

Any hints are appreciated (trying to interface a legacy paddlewheel sensor to get it on the NMEA 2000 network).

Best Regards,
Jonas

**Comment by mairas** [comment]
Hi,

I'm not sure I followed. :-) Did I get it right that you're eventually trying to interface the paddlewheel sensor but your first step is to test OPTO_IN by manually connecting GND to ISO_GND and a jumper wire from either 3.3V or GND to OPTO_IN? That should definitely work.

In particular, I verified with the following:
- Connected GND to ISO_GND
- Uploaded a minimal SensESP application with the following in `main.cpp`:

```
  pinMode(35, INPUT);
  app.onRepeat(500, [](){
      bool input = digitalRead(35);
      Serial.println(input);
  });
```

And then added another jumper cable to ISO_IN and moved that between 3.3V and GND. The printed value changes accordingly on the PlatformIO monitor.

(Note that there's nothing in my test that benefits from using SensESP - I just took an old project and used it as a basis.)

What kind of wiring does the paddlewheel sensor have? Is it an active device (having some kind of voltage input) or are there just GND and some signal wire outputs? If it's the latter, it's likely you might have trouble connecting it via an optocoupler. Optocouplers have relatively low impedance inputs, meaning that the signal source must be able to provide some current (in the range of 2 mA) to activate the optocoupler, and a passive sensor probably won't be able to do that (tried that with an inductive Yanmar RPM sensor).

**Reply by jonas1972** [reply]
Thanks for the help! You got me just right. I was trying to use the opto in as a first step. That works now. Not really sure what I did wrong, but now it works ok. As for the sensor it is this one: https://airmar.com/uploads/wiringdiagrams/91-523.pdf. I'm not sure what voltage it uses to hence the connection with the optocoupler. I tried 3.3V yesterday (without the optocoupler) but that didn't give any signal. I will try either 5V or 12V now that I got it working. As I would like to have the NMEA 2000 provide the 12V for the board (and sensor), do you see any problem in getting it from Vin Prot and feeding it through the sensor to the opto in? I have already bridged the NMEA 2000 module and the input voltage module and that part works just fine.

**Reply by mairas** [reply]
Googling for airmar 91-523 gave me this: https://www.airmartechnology.com/xref.html?refid=ST850

So apparently the sensor is just one of the million variants of ST850. The Specifications tab of the [ST850 product description page](https://www.airmartechnology.com/productdescription.html?id=103) states that the supply voltage is 9 to 16 VDC. So there you have it. :-)

As for providing an output voltage via Vin Prot, that should be fine. Just be sure to also provide GND from the same place. In that case you're of course not using the optocoupler as a galvanic isolator (which should be fine if the sensor is not galvanically connected to other circuits), and the hall effect sensor needs to be able to provide about 5 mA of current to the optocoupler input at 12V. From what I've read, they should be able to output 10-20 mA, so it'll work just fine.

If the optocoupler input is a problem you could just ditch them and clamp the output to the logic level rails with something like the following:

<img width="693" alt="Screenshot 2021-08-14 at 17 20 06" src="https://user-images.githubusercontent.com/407010/129449381-b03d79ad-cca8-4ec8-9c4c-1d0434b62f26.png">

In that circuit, the Schottky diodes clamp the input voltages to the rails. Any sensor providing a signal between 0V and some reasonably safe high logic value should work fine.
